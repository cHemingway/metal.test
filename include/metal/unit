#if defined(__cplusplus)
#include <metal/unit.hpp>
#else
#include <metal/unit.h>
#endif