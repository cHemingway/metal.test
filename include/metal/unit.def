/**
 * @file   metal/unit.def
 * @date   25.10.2016
 * @author Klemens D. Morgenstern
 *
 */

#ifndef METAL_UNIT_DEF_
#define METAL_UNIT_DEF_

typedef enum __metal_level_t
{
    __metal_level_assert,
    __metal_level_expect,
} __metal_level;

typedef enum __metal_oper_t
{
    __metal_oper_enter_case,
    __metal_oper_exit_case,
    __metal_oper_enter_ranged,
    __metal_oper_exit_ranged,
    __metal_oper_log,
    __metal_oper_checkpoint,
    __metal_oper_message,
    __metal_oper_plain,
    __metal_oper_predicate,
    __metal_oper_equal,
    __metal_oper_not_equal,
    __metal_oper_close,
    __metal_oper_close_rel,
    __metal_oper_close_perc,
    __metal_oper_ge,
    __metal_oper_greater,
    __metal_oper_le,
    __metal_oper_lesser,
    __metal_oper_exception,
    __metal_oper_any_exception,
    __metal_oper_no_exception,
    __metal_oper_no_exec,
    __metal_oper_exec,
    __metal_oper_report
} __metal_oper;

extern int __metal_status  ;
extern int __metal_critical;
extern int __metal_errored;

void __metal_call(void (*func)(), const char * msg,
               const char * file, int line);

void __metal_unit_break(__metal_level lvl,
               __metal_oper oper,
               int condition,
               int bitwise,
               const char* str1,
               const char* str2,
               const char* str3,
               const char* file,
               int line);

#if defined(METAL_UNIT_ONLY_ERRORS)
#define __metal_unit_impl(lvl, oper, condition, ...) (((condition == 0) && (lvl == __metal_level_assert)) ? __metal_unit_break(lvl, oper, condition, __VA_ARGS__) : ((void)0))
#elif defined(METAL_UNIT_ONLY_WARNINGS)
#define __metal_unit_impl(lvl, oper, condition, ...) ((condition == 0) ? __metal_unit_break(lvl, oper, condition, __VA_ARGS__) : ((void)0))
#else
#define __metal_unit_impl(...) __metal_unit_break(__VA_ARGS__)
#endif

#if defined(METAL_UNIT_ONLY_ERRORS) || defined(METAL_UNIT_ONLY_WARNINGS)
#define __metal_unit_call(Function, ...) Function()
#else
#define __metal_unit_call(...) __metal_call(__VA_ARGS__)
#endif

int __metal_report();

#define __METAL_BITWISE_STEP(Lhs, Rhs, Index, Oper) (((Lhs >> Index) & 1) Oper ((Rhs >> Index) & 1))

#define __METAL_BITWISE_8(Lhs, Rhs, Oper, Chain) \
    ((Lhs & 1) Oper (Rhs & 1))                   \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x4, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x5, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x6, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x7, Oper)

#define __METAL_BITWISE_16(Lhs, Rhs, Oper, Chain) \
    __METAL_BITWISE_8(Lhs, Rhs, Oper, Chain)        \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x8, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x9, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xA, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xB, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xC, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xD, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xE, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xF, Oper)

#define __METAL_BITWISE_32(Lhs, Rhs, Oper, Chain) \
    __METAL_BITWISE_16(Lhs, Rhs, Oper, Chain)        \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x10, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x11, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x12, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x13, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x14, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x15, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x16, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x17, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x18, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x19, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1A, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1B, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1C, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1D, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1E, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1F, Oper)

#define __METAL_BITWISE_64(Lhs, Rhs, Oper, Chain)    \
    __METAL_BITWISE_32(Lhs, Rhs, Oper, Chain)        \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x20, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x21, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x22, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x23, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x24, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x25, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x26, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x27, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x28, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x29, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2A, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2B, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2C, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2D, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2E, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2F, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x30, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x31, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x32, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x33, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x34, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x35, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x36, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x37, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x38, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x39, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3A, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3B, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3C, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3D, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3E, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3F, Oper)

#define __METAL_BITWISE_SIZE(Lhs, Rhs) ((sizeof(Lhs) >= sizeof(Rhs)) ? sizeof(Lhs) : sizeof(Rhs))

#define METAL_CONCAT_IMPL( x, y ) x##y
#define METAL_CONCAT( x, y ) METAL_CONCAT_IMPL( x, y )

#define METAL_STRINGIZE(x) METAL_STRINGIZE2(x)
#define METAL_STRINGIZE2(x) #x
#define METAL_LOCATION_STR() __FILE__ "(" METAL_STRINGIZE(__LINE__) ")"

#define __METAL_PP_NARG(...)  __METAL_PP_NARG_(__VA_ARGS__,__METAL_PP_RSEQ_N())
#define __METAL_PP_NARG_(...) __METAL_PP_ARG_N(__VA_ARGS__)
#define __METAL_PP_ARG_N(_1, _2, _3, _4, _5, _6, _7, _8, _9,_A, _B, _C, _D, _E, _F, N,...) N
#define __METAL_PP_RSEQ_N() F, E, D, C, B, A, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0

#define METAL_OVERLOAD(Macro, Args...) METAL_CONCAT(Macro, __METAL_PP_NARG(Args)) (Args)

#endif
