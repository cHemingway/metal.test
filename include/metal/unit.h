/**
 * @file   metal/test/unit.h
 * @date   10.08.2016
 * @author Klemens D. Morgenstern
 *
 *  This header provides the C++ version of the test macros.
 */
#ifndef METAL_UNIT_H_
#define METAL_UNIT_H_

#include <metal/unit.def>
#include <stdint.h>
#include <stddef.h>


#define METAL_ERROR()       !__metal_status
#define METAL_STATUS()      +__metal_status
#define METAL_ERRORED()     +__metal_errored
#define METAL_IS_CRITICAL() +__metal_critical

#define METAL_CALL(Function, Message) __metal_call(Function, Message, __FILE__, __LINE__);

#define METAL_REPORT() __metal_report()

#define METAL_BITWISE_EXPR(Lhs, Rhs, Oper, Chain)                                          \
        (__METAL_BITWISE_SIZE(Lhs, Rhs) == 1)   ? __METAL_BITWISE_8(Lhs, Rhs, Oper, Chain)  : \
        ((__METAL_BITWISE_SIZE(Lhs, Rhs) == 2)  ? __METAL_BITWISE_16(Lhs, Rhs, Oper, Chain) : \
         ((__METAL_BITWISE_SIZE(Lhs, Rhs) == 4) ? __METAL_BITWISE_32(Lhs, Rhs, Oper, Chain) : \
            __METAL_BITWISE_64(Lhs, Rhs,  Oper, Chain) ) )

#define METAL_BITWISE(Level, Lhs, Rhs, Oper, Chain, OperId) \
__metal_unit_impl(Level, OperId, METAL_BITWISE_EXPR(Lhs, Rhs, Oper, Chain), 1, #Lhs, #Rhs, #Oper, __FILE__, __LINE__);

#define METAL_RANGE_ENTER(Level, LhsDistance, RhsDistance, Message) \
__metal_unit_impl(Level, __metal_oper_enter_ranged,  LhsDistance == RhsDistance, 0, Message, #LhsDistance, #RhsDistance, __FILE__, __LINE__);

#define METAL_RANGE_EXIT(Level,  Status, Message) \
__metal_unit_impl(Level, __metal_oper_exit_ranged, Status, 0, Message, 0, 0,__FILE__, __LINE__);


#define METAL_RANGED(Level, Lhs, LhsSize, Rhs, RhsSize, MACRO )                                                             \
{                                                                                                                        \
    int status = 1;                                                                                                      \
    int i = 0;                                                                                                           \
    METAL_RANGE_ENTER(Level, LhsSize, RhsSize, "{" #Lhs "[0 : " #LhsSize "], " #Rhs "[0 : " #RhsSize "]}");                 \
    const size_t Size = LhsSize > RhsSize ? RhsSize : LhsSize;                                                           \
                                                                                                                         \
    status &= METAL_STATUS();                                                                                               \
                                                                                                                         \
    while (i < Size)                                                                                                     \
    {                                                                                                                    \
        MACRO ;                                                                                                          \
        status &= METAL_STATUS();                                                                                           \
        i++;                                                                                                             \
    }                                                                                                                    \
                                                                                                                         \
    METAL_RANGE_EXIT(Level, status, "{" #Lhs "[0 : " #LhsSize "], " #Rhs "[0 : " #RhsSize "]}");                            \
}

#define METAL_STATIC_ASSERT(Condition, Message) \
typedef char METAL_CONCAT(__metal_static_assert_, __COUNTER__) [Condition ? 1 : -1];

//Operations
#define METAL_LOG(Message) __metal_unit_impl(__metal_level_expect, __metal_oper_log, 1, 0, Message, 0, 0, __FILE__, __LINE__);
#define METAL_CHECKPOINT() __metal_unit_impl(__metal_level_expect, __metal_oper_checkpoint,        1, 0,       0, 0, 0, __FILE__, __LINE__);
#define METAL_ASSERT_MESSAGE(Condition, Message) __metal_unit_impl(__metal_level_assert, __metal_oper_message, Condition, 0, Message, 0, 0, __FILE__, __LINE__);
#define METAL_EXPECT_MESSAGE(Condition, Message) __metal_unit_impl(__metal_level_expect, __metal_oper_message, Condition, 0, Message, 0, 0, __FILE__, __LINE__);

#define METAL_ASSERT(Condition) __metal_unit_impl(__metal_level_assert, __metal_oper_plain, Condition, 0, #Condition, 0, 0, __FILE__, __LINE__);
#define METAL_EXPECT(Condition) __metal_unit_impl(__metal_level_expect, __metal_oper_plain, Condition, 0, #Condition, 0, 0, __FILE__, __LINE__);

#define METAL_ASSERT_PREDICATE(Function, Args...) __metal_unit_impl(__metal_level_assert, __metal_oper_predicate, Function(Args), 0, #Function, #Args, 0, __FILE__, __LINE__);
#define METAL_EXPECT_PREDICATE(Function, Args...) __metal_unit_impl(__metal_level_expect, __metal_oper_predicate, Function(Args), 0, #Function, #Args, 0, __FILE__, __LINE__);
#define METAL_STATIC_ASSERT_PREDICATE(Function, Args...) METAL_STATIC_ASSERT(Function(Args), #Function "(" #Args ")");

#define METAL_ASSERT_EQUAL(Lhs, Rhs) __metal_unit_impl(__metal_level_assert, __metal_oper_equal, Lhs == Rhs, 0, #Lhs, #Rhs, 0, __FILE__, __LINE__);
#define METAL_EXPECT_EQUAL(Lhs, Rhs) __metal_unit_impl(__metal_level_expect, __metal_oper_equal, Lhs == Rhs, 0, #Lhs, #Rhs, 0, __FILE__, __LINE__);
#define METAL_ASSERT_EQUAL_BITWISE(Lhs, Rhs) METAL_BITWISE(__metal_level_assert, Lhs, Rhs, ==, &&, __metal_oper_equal)
#define METAL_EXPECT_EQUAL_BITWISE(Lhs, Rhs) METAL_BITWISE(__metal_level_expect, Lhs, Rhs, ==, &&, __metal_oper_equal)

#define METAL_STATIC_ASSERT_EQUAL(Lhs, Rhs) METAL_STATIC_ASSERT(Lhs == Rhs, #Lhs " == " #Rhs)
#define METAL_STATIC_ASSERT_EQUAL_BITWISE(Lhs, Rhs) METAL_STATIC_ASSERT(METAL_BITWISE_EXPR(Rhs, Lhs, ==, &&), " [bitwise] " #Rhs " == " #Lhs)

#define METAL_ASSERT_NOT_EQUAL(Lhs, Rhs) __metal_unit_impl(__metal_level_assert, __metal_oper_not_equal, Lhs != Rhs, 0, #Lhs, #Rhs, 0, __FILE__, __LINE__);
#define METAL_EXPECT_NOT_EQUAL(Lhs, Rhs) __metal_unit_impl(__metal_level_expect, __metal_oper_not_equal, Lhs != Rhs, 0, #Lhs, #Rhs, 0, __FILE__, __LINE__);
#define METAL_ASSERT_NOT_EQUAL_BITWISE(Lhs, Rhs) METAL_BITWISE(__metal_level_assert, Lhs, Rhs, != , ||, __metal_oper_not_equal);
#define METAL_EXPECT_NOT_EQUAL_BITWISE(Lhs, Rhs) METAL_BITWISE(__metal_level_expect, Lhs, Rhs, != , ||, __metal_oper_not_equal);

#define METAL_STATIC_ASSERT_NOT_EQUAL(Lhs, Rhs) METAL_STATIC_ASSERT(Rhs != Lhs, #Lhs " != " #Rhs)
#define METAL_STATIC_ASSERT_NOT_EQUAL_BITWISE(Lhs, Rhs) METAL_STATIC_ASSERT(METAL_BITWISE_EXPR(Lhs, Rhs, !=, ||), " [bitwise] " #Lhs " != " #Rhs)

#define METAL_ASSERT_CLOSE(Lhs, Rhs, Tolerance) __metal_unit_impl(__metal_level_assert, __metal_oper_close, (Rhs <= (Lhs + Tolerance)) && (Rhs >= (Lhs - Tolerance)), 0, #Rhs, #Lhs, #Tolerance, __FILE__, __LINE__);
#define METAL_EXPECT_CLOSE(Lhs, Rhs, Tolerance) __metal_unit_impl(__metal_level_expect, __metal_oper_close, (Rhs <= (Lhs + Tolerance)) && (Rhs >= (Lhs - Tolerance)), 0, #Rhs, #Lhs, #Tolerance, __FILE__, __LINE__);
#define METAL_STATIC_ASSERT_CLOSE(Lhs, Rhs, Tolerance) \
    METAL_STATIC_ASSERT((Lhs <= (Rhs + Tolerance)) && (Lhs >= (Rhs - Tolerance)) , #Lhs " == " #Rhs " +/- " #Tolerance)

#define METAL_ASSERT_CLOSE_RELATIVE(Lhs, Rhs, Tolerance) __metal_unit_impl(__metal_level_assert, __metal_oper_close_rel, (Rhs <= (Lhs * (1. + Tolerance))) && (Rhs >= (Lhs * (1. - Tolerance))), 0, #Rhs, #Lhs, #Tolerance, __FILE__, __LINE__);
#define METAL_EXPECT_CLOSE_RELATIVE(Lhs, Rhs, Tolerance) __metal_unit_impl(__metal_level_expect, __metal_oper_close_rel, (Rhs <= (Lhs * (1. + Tolerance))) && (Rhs >= (Lhs * (1. - Tolerance))), 0, #Rhs, #Lhs, #Tolerance, __FILE__, __LINE__);
#define METAL_STATIC_ASSERT_CLOSE_RELATIVE(Lhs, Rhs, Tolerance) \
    METAL_STATIC_ASSERT((Lhs <= (Rhs * (1. + Tolerance))) && (Lhs >= (Rhs * (1. - Tolerance))) , #Lhs " == " #Rhs " +/- " #Tolerance " ~")

#define METAL_ASSERT_CLOSE_PERCENT(Lhs, Rhs, Tolerance) __metal_unit_impl(__metal_level_assert, __metal_oper_close_perc, (Rhs <= (Lhs * (1. + ( Tolerance / 100.)))) && (Rhs >= (Lhs * (1. - (Tolerance / 100.)))), 0, #Rhs, #Lhs, #Tolerance, __FILE__, __LINE__);
#define METAL_EXPECT_CLOSE_PERCENT(Lhs, Rhs, Tolerance) __metal_unit_impl(__metal_level_expect, __metal_oper_close_perc, (Rhs <= (Lhs * (1. + ( Tolerance / 100.)))) && (Rhs >= (Lhs * (1. - (Tolerance / 100.)))), 0, #Rhs, #Lhs, #Tolerance, __FILE__, __LINE__);
#define METAL_STATIC_ASSERT_CLOSE_PERCENT(Lhs, Rhs, Tolerance) \
    METAL_STATIC_ASSERT((Rhs <= (Lhs * (1. + (Tolerance / 100.)))) && (Rhs >= (Lhs * (1. - ( Tolerance / 100.)))) , #Lhs " == " #Rhs " +/- " #Tolerance "%")

#define METAL_ASSERT_GE(Lhs, Rhs) __metal_unit_impl(__metal_level_assert, __metal_oper_ge, Lhs >= Rhs, 0, #Lhs, #Rhs, 0, __FILE__, __LINE__);
#define METAL_EXPECT_GE(Lhs, Rhs) __metal_unit_impl(__metal_level_expect, __metal_oper_ge, Lhs >= Rhs, 0, #Lhs, #Rhs, 0, __FILE__, __LINE__);
#define METAL_ASSERT_GE_BITWISE(Lhs, Rhs) METAL_BITWISE(__metal_level_assert, Lhs, Rhs, >=, &&, __metal_oper_ge)
#define METAL_EXPECT_GE_BITWISE(Lhs, Rhs) METAL_BITWISE(__metal_level_expect, Lhs, Rhs, >=, &&, __metal_oper_ge)
#define METAL_STATIC_ASSERT_GE(Lhs, Rhs) METAL_STATIC_ASSERT(Lhs >= Rhs, #Lhs " >= " #Rhs)
#define METAL_STATIC_ASSERT_GE_BITWISE(Lhs, Rhs) METAL_STATIC_ASSERT(METAL_BITWISE_EXPR(Lhs, Rhs, >=, &&), " [bitwise] " #Lhs " >= " #Rhs)

#define METAL_ASSERT_LE(Lhs, Rhs) __metal_unit_impl(__metal_level_assert, __metal_oper_le, Lhs <= Rhs, 0, #Lhs, #Rhs, 0, __FILE__, __LINE__);
#define METAL_EXPECT_LE(Lhs, Rhs) __metal_unit_impl(__metal_level_expect, __metal_oper_le, Lhs <= Rhs, 0, #Lhs, #Rhs, 0, __FILE__, __LINE__);
#define METAL_ASSERT_LE_BITWISE(Lhs, Rhs) METAL_BITWISE(__metal_level_assert, Lhs, Rhs, <=, &&, __metal_oper_le)
#define METAL_EXPECT_LE_BITWISE(Lhs, Rhs) METAL_BITWISE(__metal_level_expect, Lhs, Rhs, <=, &&, __metal_oper_le)
#define METAL_STATIC_ASSERT_LE(Lhs, Rhs) METAL_STATIC_ASSERT(Lhs <= Rhs, #Lhs " <= " #Rhs)
#define METAL_STATIC_ASSERT_LE_BITWISE(Lhs, Rhs) METAL_STATIC_ASSERT(METAL_BITWISE_EXPR(Lhs, Rhs, <=, &&), " [bitwise] " #Lhs " <= " #Rhs)

#define METAL_ASSERT_GREATER(Lhs, Rhs) __metal_unit_impl(__metal_level_assert, __metal_oper_greater, Lhs > Rhs, 0, #Lhs, #Rhs, 0, __FILE__, __LINE__);
#define METAL_EXPECT_GREATER(Lhs, Rhs) __metal_unit_impl(__metal_level_expect, __metal_oper_greater, Lhs > Rhs, 0, #Lhs, #Rhs, 0, __FILE__, __LINE__);
#define METAL_STATIC_ASSERT_GREATER(Lhs, Rhs) METAL_STATIC_ASSERT(Lhs > Rhs, #Lhs " > " #Rhs)

#define METAL_ASSERT_LESSER(Lhs, Rhs) __metal_unit_impl(__metal_level_assert, __metal_oper_lesser, Lhs < Rhs, 0, #Lhs, #Rhs, 0, __FILE__, __LINE__);
#define METAL_EXPECT_LESSER(Lhs, Rhs) __metal_unit_impl(__metal_level_expect, __metal_oper_lesser, Lhs < Rhs, 0, #Lhs, #Rhs, 0, __FILE__, __LINE__);
#define METAL_STATIC_ASSERT_LESSER(Lhs, Rhs) METAL_STATIC_ASSERT(Lhs < Rhs, #Lhs " < " #Rhs)


#define METAL_ASSERT_EQUAL_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_assert, Lhs, LhsSize, Rhs, RhsSize, METAL_ASSERT_EQUAL(Lhs[i], Rhs[i]))
#define METAL_EXPECT_EQUAL_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_expect, Lhs, LhsSize, Rhs, RhsSize, METAL_EXPECT_EQUAL(Lhs[i], Rhs[i]))
#define METAL_ASSERT_EQUAL_BITWISE_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_assert, Lhs, LhsSize, Rhs, RhsSize, METAL_ASSERT_EQUAL_BITWISE(Lhs[i], Rhs[i]))
#define METAL_EXPECT_EQUAL_BITWISE_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_expect, Lhs, LhsSize, Rhs, RhsSize, METAL_EXPECT_EQUAL_BITWISE(Lhs[i], Rhs[i]))

#define METAL_ASSERT_NOT_EQUAL_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_assert, Lhs, LhsSize, Rhs, RhsSize, METAL_ASSERT_NOT_EQUAL(Lhs[i], Rhs[i]))
#define METAL_EXPECT_NOT_EQUAL_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_expect, Lhs, LhsSize, Rhs, RhsSize, METAL_EXPECT_NOT_EQUAL(Lhs[i], Rhs[i]))
#define METAL_ASSERT_NOT_EQUAL_BITWISE_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_assert, Lhs, LhsSize, Rhs, RhsSize, METAL_ASSERT_NOT_EQUAL_BITWISE(Lhs[i], Rhs[i]))
#define METAL_EXPECT_NOT_EQUAL_BITWISE_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_expect, Lhs, LhsSize, Rhs, RhsSize, METAL_EXPECT_NOT_EQUAL_BITWISE(Lhs[i], Rhs[i]))

#define METAL_ASSERT_CLOSE_RANGED(Lhs, LhsSize, Rhs, RhsSize, Tolerance) METAL_RANGED(__metal_level_assert, Lhs, LhsSize, Rhs, RhsSize, METAL_ASSERT_CLOSE(Lhs[i], Rhs[i], Tolerance))
#define METAL_EXPECT_CLOSE_RANGED(Lhs, LhsSize, Rhs, RhsSize, Tolerance) METAL_RANGED(__metal_level_expect, Lhs, LhsSize, Rhs, RhsSize, METAL_EXPECT_CLOSE(Lhs[i], Rhs[i], Tolerance))

#define METAL_ASSERT_CLOSE_RELATIVE_RANGED(Lhs, LhsSize, Rhs, RhsSize, Tolerance) METAL_RANGED(__metal_level_assert, Lhs, LhsSize, Rhs, RhsSize, METAL_ASSERT_CLOSE_RELATIVE(Lhs[i], Rhs[i], Tolerance))
#define METAL_EXPECT_CLOSE_RELATIVE_RANGED(Lhs, LhsSize, Rhs, RhsSize, Tolerance) METAL_RANGED(__metal_level_expect, Lhs, LhsSize, Rhs, RhsSize, METAL_EXPECT_CLOSE_RELATIVE(Lhs[i], Rhs[i], Tolerance))

#define METAL_ASSERT_CLOSE_PERCENT_RANGED(Lhs, LhsSize, Rhs, RhsSize, Tolerance) METAL_RANGED(__metal_level_assert, Lhs, LhsSize, Rhs, RhsSize, METAL_ASSERT_CLOSE_PERCENT(Lhs[i], Rhs[i], Tolerance))
#define METAL_EXPECT_CLOSE_PERCENT_RANGED(Lhs, LhsSize, Rhs, RhsSize, Tolerance) METAL_RANGED(__metal_level_expect, Lhs, LhsSize, Rhs, RhsSize, METAL_EXPECT_CLOSE_PERCENT(Lhs[i], Rhs[i], Tolerance))

#define METAL_ASSERT_GE_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_assert, Lhs, LhsSize, Rhs, RhsSize, METAL_ASSERT_GE(Lhs[i], Rhs[i]))
#define METAL_EXPECT_GE_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_expect, Lhs, LhsSize, Rhs, RhsSize, METAL_EXPECT_GE(Lhs[i], Rhs[i]))
#define METAL_ASSERT_GE_BITWISE_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_assert, Lhs, LhsSize, Rhs, RhsSize, METAL_ASSERT_GE_BITWISE(Lhs[i], Rhs[i]))
#define METAL_EXPECT_GE_BITWISE_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_expect, Lhs, LhsSize, Rhs, RhsSize, METAL_EXPECT_GE_BITWISE(Lhs[i], Rhs[i]))

#define METAL_ASSERT_LE_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_assert, Lhs, LhsSize, Rhs, RhsSize, METAL_ASSERT_LE(Lhs[i], Rhs[i]))
#define METAL_EXPECT_LE_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_expect, Lhs, LhsSize, Rhs, RhsSize, METAL_EXPECT_LE(Lhs[i], Rhs[i]))
#define METAL_ASSERT_LE_BITWISE_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_assert, Lhs, LhsSize, Rhs, RhsSize, METAL_ASSERT_LE_BITWISE(Lhs[i], Rhs[i]))
#define METAL_EXPECT_LE_BITWISE_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_expect, Lhs, LhsSize, Rhs, RhsSize, METAL_EXPECT_LE_BITWISE(Lhs[i], Rhs[i]))

#define METAL_ASSERT_GREATER_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_assert, Lhs, LhsSize, Rhs, RhsSize, METAL_ASSERT_GREATER(Lhs[i], Rhs[i]))
#define METAL_EXPECT_GREATER_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_expect, Lhs, LhsSize, Rhs, RhsSize, METAL_EXPECT_GREATER(Lhs[i], Rhs[i]))

#define METAL_ASSERT_LESSER_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_assert, Lhs, LhsSize, Rhs, RhsSize, METAL_ASSERT_LESSER(Lhs[i], Rhs[i]))
#define METAL_EXPECT_LESSER_RANGED(Lhs, LhsSize, Rhs, RhsSize) METAL_RANGED(__metal_level_expect, Lhs, LhsSize, Rhs, RhsSize, METAL_EXPECT_LESSER(Lhs[i], Rhs[i]))

#define METAL_ASSERT_NO_EXECUTE() __metal_unit_impl(__metal_level_assert, __metal_oper_no_exec, 0, 0, "unexpected execution", 0, 0, __FILE__, __LINE__);
#define METAL_EXPECT_NO_EXECUTE() __metal_unit_impl(__metal_level_expect, __metal_oper_no_exec, 0, 0, "unexpected execution", 0, 0, __FILE__, __LINE__);

#define METAL_ASSERT_EXECUTE() __metal_unit_impl(__metal_level_assert, __metal_oper_exec, 1, 0, "expected execution", 0, 0, __FILE__, __LINE__);
#define METAL_EXPECT_EXECUTE() __metal_unit_impl(__metal_level_expect, __metal_oper_exec, 1, 0, "expected execution", 0, 0, __FILE__, __LINE__);


#define METAL_CRITICAL(Check)  __metal_critical ++; Check ; __metal_critical--;
#define METAL_ENTER_CRITICAL() __metal_critical ++;
#define METAL_EXIT_CRITICAL()  __metal_critical --;


#if !defined (METAL_NO_IMPLEMENT_INCLUDE)
#include <metal/unit.ipp>
#endif

#endif /* METAL_TEST_UNIT_H_ */
