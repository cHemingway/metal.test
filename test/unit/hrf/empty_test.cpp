
#include <metal/unit.hpp>

int main(int argc, char * argv[])
{
    return METAL_REPORT();
}
