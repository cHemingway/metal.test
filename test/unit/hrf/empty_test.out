starting test execution 
full test report: { executed : 0, warnings : 0, errors : 0}